### ISSUE TITLE
Add Werror flag to Setup

### ISSUE THREAD

**Issue by WillAyd** [cython/cython#3482]
We recently in pandas added the -Werror flag to our setup script to help proactively identify bugs - would Cython consider doing the same thing?

**Comment by scoder** [comment]
Generally speaking: yes.

Practically speaking: the test suite isn't free of warnings currently, and getting rid of all of those isn't always easy.

We consider C compiler warnings a bug, though. Improvements welcome.

We could get closer by adding a test tag that ignores C compiler warnings for a given test, set that on all tests that currently produce them (potentially also platform specific tags), and then clean those up later. That way, we could at least enable `-Werror` for now.

**Comment by scoder** [comment]
(Let's wait with that until we fix #3481, though.)

**Comment by scoder** [comment]
Marking this as a nice beginner's ticket. What needs to be done:
- implement a new tag `ignore_cwarnings` in `runtests.py` (look for `cerror` for a similar feature)
- look through the [travis-CI](https://travis-ci.org/github/cython/cython/branches) and [appveyor](https://ci.appveyor.com/project/cython/cython/history) runs of the master branch to find tests that produce warnings.
- add the tag to these tests (`# tag: ignore_cwarnings`)
- enable `-Werror` for the test compilation (for C compilers that support it)
- see which tests fail now due to warnings and add a tag for them, too.

**Comment by WillAyd** [comment]
Taking a look at this - can you offer any insight on how to find the tests? I see this warning in the log as an example:

```sh
compiling (c/cy2) extimported ... 
=== C/C++ compiler error output: ===
cdef_syntax.c: In function â€˜__pyx_pf_11cdef_syntax_use_cdefâ€™:
cdef_syntax.c:1633:143: warning: variable â€˜__pyx_t_1â€™ set but not used [-Wunused-but-set-variable]
   __pyx_ctuple_7e355__PyObject__ptr__lParen__ptr__rParen__lParenvoid__rParen__and_PyObject__ptr__lParen__ptr__rParen__lParenvoid__rParen__etc __pyx_t_1;
                                                                                                                                               ^~~~~~~~~
====================================
```

Though when grepping for `extimported` I found nothing

**Comment by da-woods** [comment]
`./tests/compile/extimported.pyx` in that case. I did `find . -name "extimported*"`. Although that specific warning looks to come from either `./tests/compile/cdef_syntax.pyx` or `./tests/errors/cdef_syntax.pyx` - I wonder if things are being run in parallel and so the output is slightly jumbled?

**Comment by scoder** [comment]
> I wonder if things are being run in parallel and so the output is slightly jumbled?

Yes, that's the case. It usually works ok-ish to recover the failed test output, so the test turnover time is considered more important than beautiful output of the test runner. (Although the latter wouldn't hurt either…)

The warning comes from `tests/compile/cdef_syntax.pyx`, since error tests do not pass through the C compiler.

**Comment by scoder** [comment]
> warning: variable â€˜__pyx_t_1â€™ set but not used

That's definitely a bug, BTW. Temp variables should always be used, otherwise they shouldn't get created in the first place.
